Source/Core/VideoCommon/BPStructs.cpp

```diff
@@ -190,13 +192,15 @@ static void BPWritten(const BPCmd& bp, int cycles_into_future)
     }
     return;
   case BPMEM_PE_TOKEN_ID:  // Pixel Engine Token ID
+    INCSTAT(g_stats.this_frame.num_token);
```

Comment:
As an aside, any details on what a "token" is?

Reply:
See [these libogc functions](https://github.com/devkitPro/libogc/blob/3d123ebd2babebea0e59330f085fd40edb8c7233/gc/ogc/gx.h#L1658-L1692). It's used to manage synchronisation; there's a hardware register (`0xCC00100E` I think, part of the pixel engine) that indicates the current token value, and a graphics command that updates that register **when the command is processed**. So you can watch the register to know how far along processing has gotten and if you can e.g. replace data in main memory once it's no longer being used. There's also an interrupt associated with it (and actually, there are two commands for it, one for the interrupt and one without it; I'm not entirely sure how that part works or why both commands exist. `GX_SetDrawSync` uses both of them.)

For what it's worth this PR was motivated by #11098, since Super Mario Galaxy seems to use the token system before using EFB peeks for star bits. (I know TTYD also uses it to determine if the GPU has hung, though I don't fully understand the details of that part.)

Reply:
Thank you so much for the details.  11098 had me curious too (I figured this was for that).